Added core functions to allow for asynchrnous CSS loading

dainemawer opened the pull request:
<!--
### Requirements

Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change
Adds 3 new functions to `core.php` for handling asynchronous CSS mentioned in: https://github.com/10up/theme-scaffold/issues/199

The 3 functions are:
- `get_known_handles`
- `style_loader_tag`

These functions handle 3 pieces of functionality respectively:
1. Returns a collection of stylesheets we want to behave asynchronously
2. Hooks into the `style_loader_tag` action and filters the HTML that is returned by default by WordPress Core
3. When JS is disabled - the stylesheets are called normally, albeit in a manual way.

<!--
We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts.  Please include screenshots (if appropriate).
-->

### Benefits
- Selected stylesheets will now load asynchronously as the `media="print"` attribute is loaded before the DOM is complete and styles are applied. The `onload` attribute then applies the styles by updating `media="print"` to `media="all"`.

<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks

- The way the current implementation works is rather manually - my PHP has become a bit rusty over the years so I would appreciate further insight into how to get these functions to behave more dynamically.

<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process

- Stylesheets still load and apply styles when JS is disabled
- Tracked the waterfall movement of stylesheets being called in the Network tab in Chrome
- Inline JS works as styles are applied `onload`

<!--
What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., commands you ran, text you typed, buttons you clicked) and describe the results you observed.
-->

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests passed.

<!-- _NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._ -->

### Applicable Issues

<!-- Enter any applicable Issues here -->

### Changelog Entry

<!-- Add sample CHANGELOG.md entry for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related. -->

dainemawer committed:
Added core functions to allow for asynchrnous CSS loading

timwright12 commented:
going to kick this over to @christianc1

samikeijonen commented on includes/core.php:
```diff
@@ -160,6 +162,72 @@ function script_loader_tag( $tag, $handle ) {
+	// Loop over stylesheets and replace media attribute
+	foreach ( $known_handles as $known_style ) {
+		if ( $known_style === $handle ) {
+			$html = str_replace( "media='all'", "media='print' onload=\"this.media='all'\"", $html );
```
@dainemawer Would it make sense to set `noscript` in here. Something like this:

```
// Replace the media all with media print, and add an onload handler to switch it back.
$print_html = str_replace( "media='all'", "media='print' onload='this.media=\"all\"'", $html );

// Wrap the original link in a noscript tag for users who don't have js enabled.
$html = $print_html . '<noscript>' . $html . '</noscript>';
```

joesnellpdx replied:
@dainemawer Thoughts on this comment? Looks like you added it below - just making sure this is covered. If so, resolve the comment.

dainemawer replied:
I liked this idea - saves us having another function - so consider it added!

bengreeley commented on includes/core.php:
```diff
@@ -160,6 +162,72 @@ function script_loader_tag( $tag, $handle ) {
+	// Loop over all enqueued stylesheets on the page.
+	foreach ( wp_styles()->queue as $style ) {
+
+		// If we the current queue contains our known stylesheet handles append them.
```
@dainemawer Minor, but this comment seems to have a typo. Is the 'we' needed?

dainemawer replied:
@bengreeley removed this - thanks!

bengreeley commented on includes/core.php:
```diff
@@ -160,6 +162,72 @@ function script_loader_tag( $tag, $handle ) {
+
+		// If we the current queue contains our known stylesheet handles append them.
+		if ( in_array( $style, $known_handles, true ) ) {
+			$html .= "<link rel='stylesheet' href='" . esc_url( wp_styles()->registered[ $style ]->src ) . "' media='all' />" . PHP_EOL; // phpcs:ignore
```
@dainemawer The quotation marks here should be reversed. For example, the strings should be wrapped in single quotes but then the attributes should be wrapped in double-quotes. eg

`$html .= '<link rel="stylesheet" href="'`

bengreeley replied:
Actually, nevermind it looks like this is the format that WordPress outputs the markup so we should keep it consistent.

bengreeley commented on includes/core.php:
```diff
@@ -160,6 +162,72 @@ function script_loader_tag( $tag, $handle ) {
+	foreach ( wp_styles()->queue as $style ) {
+
+		// If we the current queue contains our known stylesheet handles append them.
+		if ( in_array( $style, $known_handles, true ) ) {
```
@dainemawer I'd also recommend checking whether the registered style exists before trying to access the array. Something like 

` && ! empty( wp_styles()->registered[ $style ] )`

It may be unlikely for us to encounter this scenario, but this will prevent a fatal error just in case.

dainemawer replied:
Thats a good point @bengreeley - will update!

bengreeley reviewed (approved):
@dainemawer I left a few suggestions, let me know if you have any questions

joesnellpdx commented:
@dainemawer 
I'm thinking we should move this update into the new scaffold @tlovett1 is working on and create a MR there:

https://github.com/10up/wp-scaffold

dainemawer committed:
Removed js_disabled_stylesheets, ensure filter does not run in the admin

dainemawer commented:
@joesnellpdx that sounds

> @dainemawer
> I'm thinking we should move this update into the new scaffold @tlovett1 is working on and create a MR there:
> 
> https://github.com/10up/wp-scaffold

Im good with that - let me know how to go about doing that!

dainemawer commented:
@bengreeley simplified this a bit - I removed the `js_disabled_stylesheets` function and added an `is_admin()` check as we only want this to run on the frontend. If you dont mind giving it one more eyeball - that would be awesome.

bengreeley commented:
@dainemawer Looks good

jeffpaul commented:
@joesnellpdx may be best to review open PRs in this repo, merge what's sensible, then work to port items from this repo into the new wp-scaffold one?

dainemawer commented:
@jeffpaul I like that idea - @joesnellpdx let me know if we're good to move ahead with that and I can help to port it to `wp-scaffold`

joesnellpdx commented:
> @joesnellpdx may be best to review open PRs in this repo, merge what's sensible, then work to port items from this repo into the new wp-scaffold one?

@jeffpaul That is going to be too hard to manage. Anything new, we should put into the new wp-scaffold. I'll triage the current issues and PRs in January. 

New repo: https://github.com/10up/wp-scaffold

CC: @dainemawer
